File: core/src/commonTest/kotlin/com/bluzi/babymonitor/xiaomi/SessionExpiryTest.kt
package com.bluzi.babymonitor.xiaomi

import com.bluzi.babymonitor.net.MiHttp
import com.bluzi.babymonitor.net.XiaomiSocketClosed
import com.bluzi.babymonitor.net.RawResponse
import kotlinx.coroutines.test.runTest
import kotlin.test.assertFalse
import kotlin.test.assertFailsWith
import kotlin.test.Test

// BG-8: declaring "your session expired" DELETES the stored session and forces a new sign-in.
// Getting that wrong at 3am is worse than any other bug in the app: monitoring stops and the
// parent must find their Mi password to get it back. So expiry may only be declared when Xiaomi
// itself refused the long-lived token — never when we simply could not get a straight answer.

private class ScriptedHttp(val handler: (String) -> RawResponse) : MiHttp {
    override suspend fun request(
        url: String,
        method: String,
        headers: Map<String, String>,
        body: String?,
    ): RawResponse = handler(url)
}

private fun session() = Session(
    userId = "100001",
    cUserId = "CU1",
    passToken = "PT1",
    serviceToken = "ST1",
    ssecurity = ByteArray(16),
    region = "de",
)

private fun expiredApi() = RawResponse(401, "", emptyList(), "Unauthorized".encodeToByteArray())

class SessionExpiryTest {
    private suspend fun refreshWith(http: MiHttp) = MiCloud(http, session = session()).deviceList()

    @Test
    fun `AUTH-8+BG-8 Xiaomi refusing the long-lived token is a real expiry`() = runTest {
        // The account server answers properly and hands back no location: the token is dead.
        val http = ScriptedHttp { url ->
            if (url.contains("/pass/serviceLogin?")) {
                RawResponse(200, "", emptyList(), "&&&START&&&{\"code\":70016}".encodeToByteArray())
            } else {
                expiredApi()
            }
        }
        assertFailsWith<AuthExpiredException> {
            refreshWith(http)
        }
    }

    @Test
    fun `AUTH-8+BG-8 a captive portal or maintenance page must never sign the user out`() = runTest {
        // Hotel Wi-Fi (or a Xiaomi outage) answers HTML instead of a login body. The token may be
        // perfectly good — treating this as expiry would throw away a working session.
        val http = ScriptedHttp { url ->
            if (url.contains("/pass/serviceLogin?")) {
                RawResponse(200, "", emptyList(), "<html>Sign in to WiFi</html>".encodeToByteArray())
            } else {
                expiredApi()
            }
        }
        val err = assertFailsWith<Exception> {
            refreshWith(http)
        }
        assertFalse(err is AuthExpiredException, "a non-Xiaomi response must stay retryable")
    }

    @Test
    fun `AUTH-8+BG-8 a network failure during refresh must never sign the user out`() = runTest {
        val http = ScriptedHttp { url ->
            if (url.contains("/pass/serviceLogin?")) {
                // What a real transport failure looks like on every platform: the socket layer's
                // own error, not Xiaomi's. Wi-Fi dropped, DNS failed, captive portal.
                throw XiaomiSocketClosed("http: could not resolve account.xiaomi.com")
            } else {
                expiredApi()
            }
        }
        val err = assertFailsWith<Exception> {
            refreshWith(http)
        }
        assertFalse(err is AuthExpiredException, "a transport failure must stay retryable")
    }
}

File: windows/tests/BabyMonitor.Core.Tests/SessionExpiryTest.cs
using System.Text;
using BabyMonitor.Core.Net;
using BabyMonitor.Core.Xiaomi;
using Xunit;

namespace BabyMonitor.Core.Tests;

// BG-8: declaring "your session expired" DELETES the stored session and forces a new sign-in. Getting
// that wrong at 3am is worse than any other bug in the app: monitoring stops and the parent must find
// their Mi password to get it back. So expiry may only be declared when Xiaomi itself refused the
// long-lived token — never when we simply could not get a straight answer.

internal sealed class ScriptedHttp : IMiHttp
{
    private readonly Func<string, RawResponse> _handler;

    public ScriptedHttp(Func<string, RawResponse> handler) => _handler = handler;

    public Task<RawResponse> RequestAsync(
        string url,
        string method = "GET",
        IReadOnlyDictionary<string, string>? headers = null,
        string? body = null,
        CancellationToken ct = default) => Task.FromResult(_handler(url));
}

public class SessionExpiryTest
{
    private static Session Session() => new(
        UserId: "100001",
        CUserId: "CU1",
        PassToken: "PT1",
        ServiceToken: "ST1",
        Ssecurity: new byte[16],
        Region: "de");

    private static RawResponse ExpiredApi() => new(
        401,
        string.Empty,
        Array.Empty<KeyValuePair<string, string>>(),
        Encoding.UTF8.GetBytes("Unauthorized"));

    private static Task RefreshWith(IMiHttp http) => new MiCloud(http, session: Session()).DeviceListAsync();

    [Fact(DisplayName = "AUTH-8+BG-8 Xiaomi refusing the long-lived token is a real expiry")]
    public async Task RefusedTokenIsAnExpiry()
    {
        // The account server answers properly and hands back no location: the token is dead.
        var http = new ScriptedHttp(url => url.Contains("/pass/serviceLogin?", StringComparison.Ordinal)
            ? new RawResponse(
                200,
                string.Empty,
                Array.Empty<KeyValuePair<string, string>>(),
                Encoding.UTF8.GetBytes("&&&START&&&{\"code\":70016}"))
            : ExpiredApi());

        await Assert.ThrowsAsync<AuthExpiredException>(() => RefreshWith(http));
    }

    [Fact(DisplayName = "AUTH-8+BG-8 a captive portal or maintenance page must never sign the user out")]
    public async Task CaptivePortalIsNotAnExpiry()
    {
        // Hotel Wi-Fi (or a Xiaomi outage) answers HTML instead of a login body. The token may be
        // perfectly good — treating this as expiry would throw away a working session.
        var http = new ScriptedHttp(url => url.Contains("/pass/serviceLogin?", StringComparison.Ordinal)
            ? new RawResponse(
                200,
                string.Empty,
                Array.Empty<KeyValuePair<string, string>>(),
                Encoding.UTF8.GetBytes("<html>Sign in to WiFi</html>"))
            : ExpiredApi());

        var err = await Assert.ThrowsAnyAsync<Exception>(() => RefreshWith(http));
        Assert.IsNotType<AuthExpiredException>(err);
    }

    [Fact(DisplayName = "AUTH-8+BG-8 a network failure during refresh must never sign the user out")]
    public async Task NetworkFailureIsNotAnExpiry()
    {
        var http = new ScriptedHttp(url =>
        {
            if (url.Contains("/pass/serviceLogin?", StringComparison.Ordinal))
            {
                // What a real transport failure looks like: the socket layer's own error, not
                // Xiaomi's. Wi-Fi dropped, DNS failed, captive portal.
                throw new SocketClosedException("http: could not resolve account.xiaomi.com");
            }

            return ExpiredApi();
        });

        var err = await Assert.ThrowsAnyAsync<Exception>(() => RefreshWith(http));
        Assert.IsNotType<AuthExpiredException>(err);
    }
}
